File: 백준/Silver/1874. 스택 수열/스택 수열.py
import sys

input = sys.stdin.readline
n = int(input())
stack = []
target = []
result = []
answer = []
idx = 0

for _ in range(n):
  num = int(input())
  target.append(num)

for i in range(1, n + 1):
  stack.append(i)
  answer.append('+')
  c = stack.pop()
  if c == target[idx]:
    answer.append('-')
    result.append(c)
    idx += 1
    while stack:
      if c > target[idx]:
        c = stack.pop()
        if c == target[idx]:
          answer.append('-')
          result.append(c)
          idx += 1
      else:
        break
  else:
    stack.append(i)

if result != target:
  print('NO')
else:
  print(*answer, sep='\n')

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringBuilder sb = new StringBuilder();
		int n = Integer.parseInt(br.readLine());
		
		Stack<Integer> stack = new Stack<>();
		int start = 0;

		for (int i = 0; i < n; i++) {
			int num = Integer.parseInt(br.readLine());
			if (num > start) {
				for (int j = start + 1; j <= num; j++) {
					stack.push(j);
					sb.append("+").append("\n");
				}
				start = num;
			} else if (stack.peek() != num) {
				System.out.println("NO");
				return;
			}
			
			stack.pop();
			sb.append("-").append("\n");
		}
		System.out.println(sb);
	}
}